codice/ddf
# DDF-4655 Support query mappability in WFS 1.1 source

#### What does this PR do?
Updates the 1.1.0 WfsFilterDelegate to use a MetacardMapper service to determine how to map metacard attributes to WFS feature properties. This allows users to query WFS 1.1.0 sources using DDF metacard attributes they are familiar with instead of the dynamically-generated `ext.FeatureType.FeatureProperty` attributes.

#### Who is reviewing it? 
@leo-sakh 
@Corey-Collins 
@samuelechu 
@glenhein 

#### Select relevant component teams: 
@codice/io 
@codice/ogc 

#### Ask 2 committers to review/merge the PR and tag them here.
@bdeining
@troymohl

#### How should this be tested?
1. In the Admin UI, open the Spatial app.
2. Click `Metacard to WFS Feature Map`.
3. For `Feature Type`, enter `{http://stategeothermaldata.org/uri-gin/aasg/xmlschema/welllog/0.8}AZWellLogs`.
4. For `Metacard Title to WFS Feature Property Mapping`, enter `ext.AZWellLogs.` + one of the element names from http://services.azgs.az.gov/ArcGIS/services/aasggeothermal/AZWellLogs/MapServer/WFSServer?request=DescribeFeatureType&service=WFS&version=1.1.0 that has a type of `xsd:string` such as `DisplayName` or `Notes`.
a. Do the same for `Resource URI` and `Resource Size` and try to map each to a different element.
5. For `Metacard Created Date to WFS Feature Property Mapping`, enter `ext.AZWellLogs.` + one of the element names from http://services.azgs.az.gov/ArcGIS/services/aasggeothermal/AZWellLogs/MapServer/WFSServer?request=DescribeFeatureType&service=WFS&version=1.1.0 that has a type of `xsd:dateTime` such as `StatusDate`, `SpudDate`, or `DateTimeLogRun`.
a. Do the same for `Modified`, `Effective`, and `Expiration` and try to map each to a different element.
6. For `Metacard Geography to WFS Feature Property Mapping`, enter `ext.AZWellLogs.Shape`.
7. Click `Save changes`.
8. Create a WFS 1.1.0 source with the URL http://services.azgs.az.gov/ArcGIS/services/aasggeothermal/AZWellLogs/MapServer/WFSServer
a. Disable the CN check and allow redirects. Leave everything else as-is.
8. In the karaf console, run `log:set warn org.apache.camel` and `log:set info org.apache.cxf`.
9. Open Intrigue.
10. Run queries against your WFS 1.1.0 source using the `Title`, `Created`, `Modified`, `Effective`, `Expiration`, `Resource URI`, `Resource Size`, and `Location` fields. Use different combinations of ANDs and ORs.
a. When you run a query, check the karaf console logs for a `GetFeature` XML request. Inspect the query XML and verify that the metacard attributes you queried on were appropriately mapped to their corresponding WFS features. You should not see any metacard attribute names in the query, only the WFS feature names they are mapped to.
11. Go back to `Metacard to WFS Feature Map`, edit the mappings, and click `Save changes`. Then go back to Intrigue, re-run some queries, and verify that they have the updated mappings.

#### Any background context you want to provide?
Currently, users have to use special, dynamically-created attributes to query a WFS 1.1.0 source. These attributes are of the form `ext.FeatureType.FeatureProperty` and it is not clear that these are the attributes you have to use to query the source. `anyText` and `anyGeo` are exceptions as the WFS 1.1.0 source will translate them to the appropriate feature properties understood by the WFS server and the user does not have to know about these special attributes.

#### What are the relevant tickets?
Fixes: #4655 

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
